src/lsm/lsm_tree.c

```diff
@@ -771,6 +771,7 @@ __wt_lsm_tree_switch(WT_SESSION_IMPL *session, WT_LSM_TREE *lsm_tree)
 	++lsm_tree->dsk_gen;
 
 	lsm_tree->modified = true;
+	WT_FULL_BARRIER();
```

Comment:
The barrier needs to be between the `dsk_gen` and `switch_txn` settings.  It seemed okay to put it after the `modified` line, since we need one somewhere, to make it as late so that all memory stores are visible before setting `switch_txn`.   I don't feel strongly though.